### PR TITLE
Make ** to represent zero or more folders for cloud urls

### PR DESCRIPTION
**Pull request by dilipped**
Fixes https://github.com/GoogleCloudPlatform/gsutil/issues/1276 

See b/188835024 for more info.

### CODE REVIEW

File: gslib/tests/test_wildcard_iterator.py [hunk 1]
```diff
@@ -284,6 +306,64 @@ def testDoesNotStripDelimiterForDoubleWildcard(self):
     self.assertEqual(expected_prefixes, actual_prefixes)
     self.assertEqual(expected_uri_strs, actual_uri_strs)
 
+  def testDoubleWildcardAfterBucket(self):
+    """Tests gs://bucket/**/object matching."""
+    actual_uri_strs = set()
+    actual_prefixes = set()
+    for blr in self._test_wildcard_iterator(
```
Comment (NickGoog) [inline gslib/tests/test_wildcard_iterator.py:313]: What is "blr"? Is there a more clear name?

File: gslib/tests/test_wildcard_iterator.py [hunk 2]
```diff
@@ -284,6 +306,64 @@ def testDoesNotStripDelimiterForDoubleWildcard(self):
...
+    expected_uri_strs = set([
+        self.test_bucket2_uri.clone_replace_name('double/f.txt').uri,
+        self.test_bucket2_uri.clone_replace_name('double/foo/f.txt').uri,
+        self.test_bucket2_uri.clone_replace_name('double/bar/f.txt').uri,
+    ])
+    expected_prefixes = set()
+    self.assertEqual(expected_prefixes, actual_prefixes)
+    self.assertEqual(expected_uri_strs, actual_uri_strs)
+
+  def testDoubleWildcardAfterBucketAndPrefix(self):
```
Comment (NickGoog) [inline gslib/tests/test_wildcard_iterator.py:348]: "BeforeAndAfterPrefix"?

File: gslib/wildcard_iterator.py [hunk 3]
```diff
@@ -221,7 +221,8 @@ def __iter__(self,
           url = StorageUrlFromString(urls_needing_expansion.pop(0))
           (prefix, delimiter, prefix_wildcard,
            suffix_wildcard) = (self._BuildBucketFilterStrings(url.object_name))
-          prog = re.compile(fnmatch.translate(prefix_wildcard))
+          # prog = re.compile(fnmatch.translate(prefix_wildcard))
```
Comment (NickGoog) [inline gslib/wildcard_iterator.py:224]: Left over comment

File: gslib/wildcard_iterator.py [hunk 4]
```diff
@@ -235,44 +236,81 @@ def __iter__(self,
...
-                if suffix_wildcard and rstripped_prefix != suffix_wildcard:
-                  # There's more wildcard left to expand.
-                  url_append_string = '%s%s' % (bucket_url_string,
-                                                rstripped_prefix + '/' +
-                                                suffix_wildcard)
-                  urls_needing_expansion.append(url_append_string)
-                else:
-                  # No wildcard to expand, just yield the prefix
-                  yield self._GetPrefixRef(bucket_url_string, prefix)
+            for prog in regex_patterns:
```
Comment (NickGoog) [inline gslib/wildcard_iterator.py:239]: Can we use a more clear variable name than "prog"? Maybe just "pattern"

File: gslib/wildcard_iterator.py [hunk 5]
```diff
@@ -235,44 +236,81 @@ def __iter__(self,
...
+                rstripped_prefix = StripOneSlash(prefix)
+                if prog.match(rstripped_prefix):
+                  if suffix_wildcard and rstripped_prefix != suffix_wildcard:
+                    # There's more wildcard left to expand.
+                    url_append_string = '%s%s' % (bucket_url_string,
+                                                  rstripped_prefix + '/' +
+                                                  suffix_wildcard)
+                    urls_needing_expansion.append(url_append_string)
+                  else:
+                    # No wildcard to expand, just yield the prefix
```
Comment (NickGoog) [inline gslib/wildcard_iterator.py:278]: I'm guessing a lot of this isn't your code, or I would review more closely.
But can you please add a period here? 😄

**Review by NickGoog — APPROVED** [review]

**Comment by NickGoog** [comment]
Don't merge until Kokoro passes, please!

File: gslib/tests/test_wildcard_iterator.py (hunk 1 above)
Comment (dilipped) [reply gslib/tests/test_wildcard_iterator.py:313]: It stands for BucketListingRef. I know it's not a great name, but is used widely in gsutil, so I will leave it like that for now.

File: gslib/tests/test_wildcard_iterator.py (hunk 2 above)
Comment (dilipped) [reply gslib/tests/test_wildcard_iterator.py]: I like it, done.

File: gslib/wildcard_iterator.py (hunk 3 above)
Comment (dilipped) [reply gslib/wildcard_iterator.py]: Oops

File: gslib/wildcard_iterator.py (hunk 4 above)
Comment (dilipped) [reply gslib/wildcard_iterator.py]: Done.

File: gslib/wildcard_iterator.py (hunk 5 above)
Comment (dilipped) [reply gslib/wildcard_iterator.py]: Done.

**Pull request merged**